File: code/packages/haskell/wasm-module-encoder/src/WasmModuleEncoder.hs
module WasmModuleEncoder
    ( description
    , wasmMagic
    , wasmVersion
    , WasmEncodeError(..)
    , encodeModule
    ) where

import qualified Data.ByteString as BS
import Data.ByteString (ByteString)
import qualified Data.List as List
import Data.Word (Word8)
import WasmLeb128 (encodeUnsigned)
import WasmTypes hiding (description)

description :: String
description = "Haskell WebAssembly binary encoder for the core module format"

wasmMagic :: ByteString
wasmMagic = BS.pack [0x00, 0x61, 0x73, 0x6D]

wasmVersion :: ByteString
wasmVersion = BS.pack [0x01, 0x00, 0x00, 0x00]

data WasmEncodeError = WasmEncodeError
    { wasmEncodeErrorMessage :: String
    }
    deriving (Eq, Show)

encodeModule :: WasmModule -> Either WasmEncodeError ByteString
encodeModule moduleValue = do
    customSections <- mapM (encodeSection 0 . encodeCustomSection) (wasmCustomSections moduleValue)
    typeSection <- encodeOptionalSection 1 (wasmTypes moduleValue) encodeFuncType
    importSection <- encodeOptionalSection 2 (wasmImports moduleValue) encodeImport
    functionSection <- encodeOptionalSection 3 (wasmFunctions moduleValue) (Right . encodeUnsigned)
    tableSection <- encodeOptionalSection 4 (wasmTables moduleValue) (Right . encodeTableType)
    memorySection <- encodeOptionalSection 5 (wasmMemories moduleValue) (Right . encodeMemoryType)
    globalSection <- encodeOptionalSection 6 (wasmGlobals moduleValue) (Right . encodeGlobal)
    exportSection <- encodeOptionalSection 7 (wasmExports moduleValue) (Right . encodeExport)
    startSection <-
        case wasmStart moduleValue of
            Nothing -> Right BS.empty
            Just startIndex -> encodeSection 8 (Right (encodeUnsigned startIndex))
    elementSection <- encodeOptionalSection 9 (wasmElements moduleValue) (Right . encodeElement)
    codeSection <- encodeOptionalSection 10 (wasmCode moduleValue) (Right . encodeFunctionBody)
    dataSection <- encodeOptionalSection 11 (wasmDataSegments moduleValue) (Right . encodeDataSegment)
    Right
        ( BS.concat
            [ wasmMagic
            , wasmVersion
            , BS.concat customSections
            , typeSection
            , importSection
            , functionSection
            , tableSection
            , memorySection
            , globalSection
            , exportSection
            , startSection
            , elementSection
            , codeSection
            , dataSection
            ]
        )

encodeOptionalSection :: Word8 -> [a] -> (a -> Either WasmEncodeError ByteString) -> Either WasmEncodeError ByteString
encodeOptionalSection _ [] _ = Right BS.empty
encodeOptionalSection sectionId values encoder =
    encodeSection sectionId (encodeVector values encoder)

encodeSection :: Word8 -> Either WasmEncodeError ByteString -> Either WasmEncodeError ByteString
encodeSection sectionId payloadResult = do
    payload <- payloadResult
    Right (BS.cons sectionId (encodeUnsigned (toInteger (BS.length payload)) <> payload))

encodeVector :: [a] -> (a -> Either WasmEncodeError ByteString) -> Either WasmEncodeError ByteString
encodeVector values encoder = do
    encodedValues <- mapM encoder values
    Right (encodeUnsigned (toInteger (length values)) <> BS.concat encodedValues)

encodeName :: String -> ByteString
encodeName text =
    let bytesValue = BS.pack (map (fromIntegral . fromEnum) text)
     in encodeUnsigned (toInteger (BS.length bytesValue)) <> bytesValue

encodeValueTypes :: [ValueType] -> ByteString
encodeValueTypes values = encodeUnsigned (toInteger (length values)) <> BS.pack (map valueTypeByte values)

encodeFuncType :: FuncType -> Either WasmEncodeError ByteString
encodeFuncType functionType =
    Right (BS.pack [0x60] <> encodeValueTypes (funcTypeParams functionType) <> encodeValueTypes (funcTypeResults functionType))

encodeLimits :: Limits -> ByteString
encodeLimits limitsValue =
    case limitsMax limitsValue of
        Nothing -> BS.pack [0x00] <> encodeUnsigned (limitsMin limitsValue)
        Just maximumValue -> BS.pack [0x01] <> encodeUnsigned (limitsMin limitsValue) <> encodeUnsigned maximumValue

encodeMemoryType :: MemoryType -> ByteString
encodeMemoryType = encodeLimits . memoryTypeLimits

encodeTableType :: TableType -> ByteString
encodeTableType tableType =
    BS.pack [tableTypeElementType tableType] <> encodeLimits (tableTypeLimits tableType)

encodeGlobalType :: GlobalType -> ByteString
encodeGlobalType globalTypeValue =
    BS.pack [valueTypeByte (globalValueType globalTypeValue), if globalMutable globalTypeValue then 0x01 else 0x00]

encodeImport :: Import -> Either WasmEncodeError ByteString
encodeImport importValue =
    case (importKind importValue, importTypeInfo importValue) of
        (ExternalFunction, ImportFunctionType typeIndex) ->
            Right
                ( encodeName (importModuleName importValue)
                    <> encodeName (importName importValue)
                    <> BS.pack [externalKindByte ExternalFunction]
                    <> encodeUnsigned typeIndex
                )
        (ExternalTable, ImportTableType tableType) ->
            Right
                ( encodeName (importModuleName importValue)
                    <> encodeName (importName importValue)
                    <> BS.pack [externalKindByte ExternalTable]
                    <> encodeTableType tableType
                )
        (ExternalMemory, ImportMemoryType memoryType) ->
            Right
                ( encodeName (importModuleName importValue)
                    <> encodeName (importName importValue)
                    <> BS.pack [externalKindByte ExternalMemory]
                    <> encodeMemoryType memoryType
                )
        (ExternalGlobal, ImportGlobalType globalTypeValue) ->
            Right
                ( encodeName (importModuleName importValue)
                    <> encodeName (importName importValue)
                    <> BS.pack [externalKindByte ExternalGlobal]
                    <> encodeGlobalType globalTypeValue
                )
        _ ->
            Left
                WasmEncodeError
                    { wasmEncodeErrorMessage = "import kind does not match import metadata"
                    }

encodeExport :: Export -> ByteString
encodeExport exportValue =
    encodeName (exportName exportValue)
        <> BS.pack [externalKindByte (exportKind exportValue)]
        <> encodeUnsigned (exportIndex exportValue)

encodeGlobal :: Global -> ByteString
encodeGlobal globalValue = encodeGlobalType (globalType globalValue) <> globalInitExpr globalValue

encodeElement :: Element -> ByteString
encodeElement elementValue =
    encodeUnsigned (elementTableIndex elementValue)
        <> elementOffsetExpr elementValue
        <> encodeUnsigned (toInteger (length (elementFunctionIndices elementValue)))
        <> BS.concat (map encodeUnsigned (elementFunctionIndices elementValue))

encodeDataSegment :: DataSegment -> ByteString
encodeDataSegment segment =
    encodeUnsigned (dataSegmentMemoryIndex segment)
        <> dataSegmentOffsetExpr segment
        <> encodeUnsigned (toInteger (BS.length (dataSegmentBytes segment)))
        <> dataSegmentBytes segment

encodeFunctionBody :: FunctionBody -> ByteString
encodeFunctionBody body =
    let localGroups = groupLocals (functionBodyLocals body)
        localPayload =
            encodeUnsigned (toInteger (length localGroups))
                <> BS.concat
                    [ encodeUnsigned (toInteger count) <> BS.pack [valueTypeByte valueType]
                    | (count, valueType) <- localGroups
                    ]
        payload = localPayload <> functionBodyCode body
     in encodeUnsigned (toInteger (BS.length payload)) <> payload

groupLocals :: [ValueType] -> [(Int, ValueType)]
groupLocals [] = []
groupLocals values =
    map toPair (List.group values)
  where
    toPair groupValues =
        case groupValues of
            [] -> error "groupLocals received an empty group"
            valueType : _ -> (length groupValues, valueType)

encodeCustomSection :: CustomSection -> Either WasmEncodeError ByteString
encodeCustomSection customSection =
    Right (encodeName (customSectionName customSection) <> customSectionData customSection)

File: code/packages/elixir/wasm_module_encoder/lib/coding_adventures/wasm_module_encoder.ex
defmodule CodingAdventures.WasmModuleEncoder.WasmEncodeError do
  defexception [:message]
end

defmodule CodingAdventures.WasmModuleEncoder do
  @moduledoc """
  Encode a `CodingAdventures.WasmTypes.WasmModule` into raw WebAssembly bytes.
  """

  alias CodingAdventures.WasmLeb128
  alias CodingAdventures.WasmModuleEncoder.WasmEncodeError
  alias CodingAdventures.WasmTypes

  alias CodingAdventures.WasmTypes.{
    CustomSection,
    DataSegment,
    Element,
    Export,
    FuncType,
    FunctionBody,
    Global,
    GlobalType,
    Import,
    Limits,
    MemoryType,
    TableType,
    WasmModule
  }

  @wasm_magic <<0x00, 0x61, 0x73, 0x6D>>
  @wasm_version <<0x01, 0x00, 0x00, 0x00>>

  @spec encode_module(WasmModule.t()) :: binary()
  def encode_module(%WasmModule{} = module) do
    sections =
      []
      |> maybe_custom_sections(module.customs)
      |> maybe_section(1, module.types, &encode_func_type/1)
      |> maybe_section(2, module.imports, &encode_import/1)
      |> maybe_section(3, module.functions, &u32/1)
      |> maybe_section(4, module.tables, &encode_table_type/1)
      |> maybe_section(5, module.memories, &encode_memory_type/1)
      |> maybe_section(6, module.globals, &encode_global/1)
      |> maybe_section(7, module.exports, &encode_export/1)
      |> maybe_start_section(module.start)
      |> maybe_section(9, module.elements, &encode_element/1)
      |> maybe_section(10, module.code, &encode_function_body/1)
      |> maybe_section(11, module.data, &encode_data_segment/1)

    IO.iodata_to_binary([@wasm_magic, @wasm_version | sections])
  end

  defp maybe_custom_sections(acc, customs) do
    Enum.reduce(customs, acc, fn custom, acc_sections ->
      acc_sections ++ [section(0, encode_custom(custom))]
    end)
  end

  defp maybe_section(acc, _id, [], _encoder), do: acc

  defp maybe_section(acc, id, values, encoder) do
    acc ++ [section(id, vector(values, encoder))]
  end

  defp maybe_start_section(acc, nil), do: acc
  defp maybe_start_section(acc, start_index), do: acc ++ [section(8, u32(start_index))]

  defp section(section_id, payload) do
    [<<section_id>>, u32(IO.iodata_length(payload)), payload]
  end

  defp u32(value), do: WasmLeb128.encode_unsigned(value)

  defp name(text) do
    data = :unicode.characters_to_binary(text, :utf8)
    [u32(byte_size(data)), data]
  end

  defp vector(values, encoder) do
    encoded = Enum.map(values, encoder)
    [u32(length(values)) | encoded]
  end

  defp encode_value_types(types) do
    [u32(length(types)), Enum.map(types, &encode_value_type/1)]
  end

  defp encode_func_type(%FuncType{params: params, results: results}) do
    [<<0x60>>, encode_value_types(params), encode_value_types(results)]
  end

  defp encode_limits(%Limits{min: min, max: nil}) do
    [<<0x00>>, u32(min)]
  end

  defp encode_limits(%Limits{min: min, max: max}) do
    [<<0x01>>, u32(min), u32(max)]
  end

  defp encode_memory_type(%MemoryType{limits: limits}), do: encode_limits(limits)

  defp encode_table_type(%TableType{element_type: element_type, limits: limits}) do
    [<<element_type>>, encode_limits(limits)]
  end

  defp encode_global_type(%GlobalType{value_type: value_type, mutable: mutable}) do
    [<<encode_value_type(value_type), if(mutable, do: 0x01, else: 0x00)>>]
  end

  defp encode_import(%Import{} = import) do
    payload =
      case {import.kind, import.type_info} do
        {:function, {:function, type_idx}} ->
          u32(type_idx)

        {:function, type_idx} when is_integer(type_idx) ->
          u32(type_idx)

        {:table, {:table, %TableType{} = table_type}} ->
          encode_table_type(table_type)

        {:table, %TableType{} = table_type} ->
          encode_table_type(table_type)

        {:memory, {:memory, %MemoryType{} = memory_type}} ->
          encode_memory_type(memory_type)

        {:memory, %MemoryType{} = memory_type} ->
          encode_memory_type(memory_type)

        {:global, {:global, %GlobalType{} = global_type}} ->
          encode_global_type(global_type)

        {:global, %GlobalType{} = global_type} ->
          encode_global_type(global_type)

        {:function, other} ->
          raise WasmEncodeError,
            message: "function imports require a type index, got: #{inspect(other)}"

        {:table, other} ->
          raise WasmEncodeError,
            message: "table imports require TableType metadata, got: #{inspect(other)}"

        {:memory, other} ->
          raise WasmEncodeError,
            message: "memory imports require MemoryType metadata, got: #{inspect(other)}"

        {:global, other} ->
          raise WasmEncodeError,
            message: "global imports require GlobalType metadata, got: #{inspect(other)}"

        {kind, _other} ->
          raise WasmEncodeError, message: "unsupported import kind: #{inspect(kind)}"
      end

    [
      name(import.module_name),
      name(import.name),
      <<encode_external_kind(import.kind)>>,
      payload
    ]
  end

  defp encode_export(%Export{name: name, kind: kind, index: index}) do
    [name(name), <<encode_external_kind(kind)>>, u32(index)]
  end

  defp encode_global(%Global{global_type: global_type, init_expr: init_expr}) do
    [encode_global_type(global_type), init_expr]
  end

  defp encode_element(%Element{
         table_index: table_index,
         offset_expr: offset_expr,
         function_indices: indices
       }) do
    [u32(table_index), offset_expr, u32(length(indices)), Enum.map(indices, &u32/1)]
  end

  defp encode_data_segment(%DataSegment{
         memory_index: memory_index,
         offset_expr: offset_expr,
         data: data
       }) do
    [u32(memory_index), offset_expr, u32(byte_size(data)), data]
  end

  defp encode_function_body(%FunctionBody{locals: locals, code: code}) do
    local_groups = group_locals(locals)

    payload = [
      u32(length(local_groups)),
      Enum.map(local_groups, fn {count, value_type} ->
        [u32(count), <<encode_value_type(value_type)>>]
      end),
      code
    ]

    [u32(IO.iodata_length(payload)), payload]
  end

  defp group_locals([]), do: []

  defp group_locals([first | rest]) do
    {groups, count, current_type} =
      Enum.reduce(rest, {[], 1, first}, fn value_type, {groups, count, current_type} ->
        if value_type == current_type do
          {groups, count + 1, current_type}
        else
          {[{count, current_type} | groups], 1, value_type}
        end
      end)

    Enum.reverse([{count, current_type} | groups])
  end

  defp encode_custom(%CustomSection{name: name, data: data}) do
    [name(name), data]
  end

  defp encode_value_type(type), do: WasmTypes.value_type(type)

  defp encode_external_kind(kind), do: WasmTypes.external_kind(kind)
end
